File: examples/typescript/servers/advanced/all_networks.ts
/**
 * All Networks Server Example
 *
 * Demonstrates how to create a server that supports all available networks with
 * optional chain configuration via environment variables.
 *
 * New chain support should be added here in alphabetic order by network prefix
 * (e.g., "algorand" before "aptos" before "ccd" before "eip155" before "hedera" before "near" before "solana" before "stellar" before "tvm" before "xrpl").
 */

import { config } from "dotenv";
import express from "express";
import { paymentMiddleware, x402ResourceServer } from "@x402/express";
import { APTOS_TESTNET_CAIP2 } from "@x402/aptos";
import { ExactAptosScheme } from "@x402/aptos/exact/server";
import { ExactAvmScheme } from "@x402/avm/exact/server";
import { ALGORAND_TESTNET_CAIP2 } from "@x402/avm";
import { ExactConcordiumScheme } from "@x402/concordium/exact/server";
import { ExactEvmScheme } from "@x402/evm/exact/server";
import { ExactHederaScheme } from "@x402/hedera/exact/server";
import { ExactSvmScheme } from "@x402/svm/exact/server";
import { KEETA_TESTNET_CAIP2 } from "@x402/keeta";
import { ExactKeetaScheme } from "@x402/keeta/exact/server";
import { NEAR_TESTNET_CAIP2 } from "@x402/near";
import { ExactNearScheme } from "@x402/near/exact/server";
import { ExactStellarScheme } from "@x402/stellar/exact/server";
import { ExactTvmScheme } from "@x402/tvm/exact/server";
import { XRPL_TESTNET } from "@x402/xrpl";
import { ExactXrplScheme } from "@x402/xrpl/exact/server";
import { HTTPFacilitatorClient } from "@x402/core/server";
import type { Network, Price } from "@x402/core/types";

config();

// Configuration - optional per network
const avmAddress = process.env.AVM_ADDRESS as string | undefined;
const aptosAddress = process.env.APTOS_ADDRESS as string | undefined;
const ccdAddress = process.env.CCD_ADDRESS as string | undefined;
const evmAddress = process.env.EVM_ADDRESS as `0x${string}` | undefined;
const hederaAddress = process.env.HEDERA_ACCOUNT_ID as string | undefined;
const keetaAddress = process.env.KEETA_ADDRESS as string | undefined;
const nearAddress = process.env.NEAR_ADDRESS as string | undefined;
const svmAddress = process.env.SVM_ADDRESS as string | undefined;
const stellarAddress = process.env.STELLAR_ADDRESS as string | undefined;
const tvmAddress = process.env.TVM_ADDRESS as string | undefined;
const xrplAddress = process.env.XRPL_ADDRESS as string | undefined;

// Validate at least one address is provided
if (
  !avmAddress &&
  !aptosAddress &&
  !ccdAddress &&
  !evmAddress &&
  !svmAddress &&
  !keetaAddress &&
  !nearAddress &&
  !stellarAddress &&
  !hederaAddress &&
  !tvmAddress &&
  !xrplAddress
) {
  console.error(
    "❌ At least one of AVM_ADDRESS, APTOS_ADDRESS, CCD_ADDRESS, EVM_ADDRESS, KEETA_ADDRESS, NEAR_ADDRESS, SVM_ADDRESS, STELLAR_ADDRESS, HEDERA_ACCOUNT_ID, TVM_ADDRESS, or XRPL_ADDRESS is required",
  );
  process.exit(1);
}

const facilitatorUrl = process.env.FACILITATOR_URL;
if (!facilitatorUrl) {
  console.error("❌ FACILITATOR_URL environment variable is required");
  process.exit(1);
}

// Network configuration
const AVM_NETWORK = (process.env.AVM_NETWORK || ALGORAND_TESTNET_CAIP2) as Network; // Algorand Testnet
const APTOS_NETWORK = (process.env.APTOS_NETWORK || APTOS_TESTNET_CAIP2) as Network; // Aptos Testnet
const CCD_NETWORK = "ccd:4221332d34e1694168c2a0c0b3fd0f27" as const; // Concordium Testnet
const EVM_NETWORK = "eip155:84532" as const; // Base Sepolia
const HEDERA_NETWORK = "hedera:testnet" as const; // Hedera Testnet
const KEETA_NETWORK = KEETA_TESTNET_CAIP2; // Keeta Testnet
const NEAR_NETWORK = (process.env.NEAR_NETWORK || NEAR_TESTNET_CAIP2) as Network; // NEAR Testnet
const SVM_NETWORK = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1" as const; // Solana Devnet
const STELLAR_NETWORK = "stellar:testnet" as const; // Stellar Testnet
const HEDERA_HBAR_ASSET = "0.0.0" as const; // Native HBAR asset id
const HEDERA_WEATHER_PRICE_TINYBARS = "100000" as const; // 0.001 HBAR
const TVM_NETWORK = (process.env.TVM_NETWORK || "tvm:-3") as Network; // TON Testnet
const XRPL_NETWORK = (process.env.XRPL_NETWORK || XRPL_TESTNET) as Network; // XRPL Testnet
const CCD_WEATHER_PRICE_MICRO_CCD = "1000" as const; // 0.001 CCD

// Build accepts array dynamically based on configured addresses
const accepts: Array<{
  scheme: string;
  price: Price;
  network: Network;
  payTo: string;
}> = [];
if (avmAddress) {
  accepts.push({
    scheme: "exact",
    price: "$0.001",
    network: AVM_NETWORK,
    payTo: avmAddress,
  });
}
if (aptosAddress) {
  accepts.push({
    scheme: "exact",
    price: "$0.001",
    network: APTOS_NETWORK,
    payTo: aptosAddress,
  });
}
if (ccdAddress) {
  accepts.push({
    scheme: "exact",
    price: {
      amount: CCD_WEATHER_PRICE_MICRO_CCD,
      asset: "CCD",
    },
    network: CCD_NETWORK,
    payTo: ccdAddress,
  });
}
if (evmAddress) {
  accepts.push({
    scheme: "exact",
    price: "$0.001",
    network: EVM_NETWORK,
    payTo: evmAddress,
  });
}
if (hederaAddress) {
  accepts.push({
    scheme: "exact",
    price: {
      amount: HEDERA_WEATHER_PRICE_TINYBARS,
      asset: HEDERA_HBAR_ASSET,
    },
    network: HEDERA_NETWORK,
    payTo: hederaAddress,
  });
}
if (keetaAddress) {
  accepts.push({
    scheme: "exact",
    price: "$0.001",
    network: KEETA_NETWORK,
    payTo: keetaAddress,
  });
}
if (nearAddress) {
  accepts.push({
    scheme: "exact",
    price: "$0.001",
    network: NEAR_NETWORK,
    payTo: nearAddress,
  });
}
if (svmAddress) {
  accepts.push({
    scheme: "exact",
    price: "$0.001",
    network: SVM_NETWORK,
    payTo: svmAddress,
  });
}
if (stellarAddress) {
  accepts.push({
    scheme: "exact",
    price: "$0.001",
    network: STELLAR_NETWORK,
    payTo: stellarAddress,
  });
}
if (tvmAddress) {
  accepts.push({
    scheme: "exact",
    price: "$0.001",
    network: TVM_NETWORK,
    payTo: tvmAddress,
  });
}
if (xrplAddress) {
  accepts.push({
    scheme: "exact",
    price: {
      amount: process.env.XRPL_AMOUNT || "1000",
      asset: "XRP",
    },
    network: XRPL_NETWORK,
    payTo: xrplAddress,
  });
}

// Create facilitator client
const facilitatorClient = new HTTPFacilitatorClient({ url: facilitatorUrl });

// Create x402 resource server and register schemes dynamically
const server = new x402ResourceServer(facilitatorClient);
if (avmAddress) {
  server.register(AVM_NETWORK, new ExactAvmScheme());
}
if (aptosAddress) {
  server.register(APTOS_NETWORK, new ExactAptosScheme());
}
if (ccdAddress) {
  server.register(CCD_NETWORK, new ExactConcordiumScheme());
}
if (evmAddress) {
  server.register(EVM_NETWORK, new ExactEvmScheme());
}
if (hederaAddress) {
  server.register(HEDERA_NETWORK, new ExactHederaScheme());
}
if (keetaAddress) {
  server.register(KEETA_NETWORK, new ExactKeetaScheme());
}
if (nearAddress) {
  server.register(NEAR_NETWORK, new ExactNearScheme());
}
if (svmAddress) {
  server.register(SVM_NETWORK, new ExactSvmScheme());
}
if (stellarAddress) {
  server.register(STELLAR_NETWORK, new ExactStellarScheme());
}
if (tvmAddress) {
  server.register(TVM_NETWORK, new ExactTvmScheme());
}
if (xrplAddress) {
  server.register(XRPL_NETWORK, new ExactXrplScheme());
}

// Create Express app
const app = express();

// Apply payment middleware
app.use(
  paymentMiddleware(
    {
      "GET /weather": {
        accepts,
        description: "Weather data",
        mimeType: "application/json",
      },
    },
    server,
  ),
);

// Protected endpoint
app.get("/weather", (req, res) => {
  res.send({
    report: {
      weather: "sunny",
      temperature: 70,
    },
  });
});

// Health check (no payment required)
app.get("/health", (req, res) => {
  res.send({ status: "ok" });
});

// Start server
const port = process.env.PORT || 4021;
app.listen(port, () => {
  console.log(`🚀 All Networks Server listening at http://localhost:${port}`);
  if (avmAddress) {
    console.log(`   AVM: ${avmAddress} on ${AVM_NETWORK}`);
  }
  if (aptosAddress) {
    console.log(`   Aptos: ${aptosAddress} on ${APTOS_NETWORK}`);
  }
  if (ccdAddress) {
    console.log(`   CCD: ${ccdAddress} on ${CCD_NETWORK}`);
  }
  if (evmAddress) {
    console.log(`   EVM: ${evmAddress} on ${EVM_NETWORK}`);
  }
  if (hederaAddress) {
    console.log(`   Hedera: ${hederaAddress} on ${HEDERA_NETWORK}`);
  }
  if (keetaAddress) {
    console.log(`   Keeta: ${keetaAddress} on ${KEETA_NETWORK}`);
  }
  if (nearAddress) {
    console.log(`   NEAR: ${nearAddress} on ${NEAR_NETWORK}`);
  }
  if (svmAddress) {
    console.log(`   SVM: ${svmAddress} on ${SVM_NETWORK}`);
  }
  if (stellarAddress) {
    console.log(`   Stellar: ${stellarAddress} on ${STELLAR_NETWORK}`);
  }
  if (tvmAddress) {
    console.log(`   TVM: ${tvmAddress} on ${TVM_NETWORK}`);
  }
  if (xrplAddress) {
    console.log(`   XRPL: ${xrplAddress} on ${XRPL_NETWORK}`);
  }
  console.log(`   Facilitator: ${facilitatorUrl}`);
  console.log();
});

File: examples/python/servers/advanced/all_networks.py
"""All Networks Server Example.

Demonstrates how to create a server that supports all available networks with
optional chain configuration via environment variables.

New chain support should be added here in alphabetic order by network prefix
(e.g., "cardano" before "eip155" before "solana" before "tvm").
"""

import os
import sys

from dotenv import load_dotenv
from fastapi import FastAPI
from pydantic import BaseModel

from x402.http import FacilitatorConfig, HTTPFacilitatorClient, PaymentOption
from x402.http.middleware.fastapi import PaymentMiddlewareASGI
from x402.http.types import RouteConfig
from x402.mechanisms.cardano.exact import ExactCardanoServerScheme
from x402.mechanisms.evm.exact import ExactEvmServerScheme
from x402.mechanisms.svm.exact import ExactSvmServerScheme
from x402.mechanisms.tvm import TVM_TESTNET
from x402.mechanisms.tvm.exact import ExactTvmServerScheme
from x402.schemas import Network
from x402.server import x402ResourceServer

load_dotenv()

# Configuration - optional per network
CARDANO_ADDRESS = os.getenv("CARDANO_ADDRESS")
EVM_ADDRESS = os.getenv("EVM_ADDRESS")
SVM_ADDRESS = os.getenv("SVM_ADDRESS")
TVM_ADDRESS = os.getenv("TVM_ADDRESS")

# Validate at least one address is provided
if not CARDANO_ADDRESS and not EVM_ADDRESS and not SVM_ADDRESS and not TVM_ADDRESS:
    print("❌ Configure CARDANO_ADDRESS, EVM_ADDRESS, SVM_ADDRESS, or TVM_ADDRESS")
    sys.exit(1)

# Network configuration
CARDANO_NETWORK: Network = os.getenv("CARDANO_NETWORK", "cardano:preprod")
EVM_NETWORK: Network = "eip155:84532"  # Base Sepolia
SVM_NETWORK: Network = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1"  # Solana Devnet
TVM_NETWORK: Network = os.getenv("TVM_NETWORK", TVM_TESTNET)  # TON testnet by default
FACILITATOR_URL = os.getenv("FACILITATOR_URL", "https://x402.org/facilitator")


# Response schemas
class WeatherReport(BaseModel):
    weather: str
    temperature: int


class WeatherResponse(BaseModel):
    report: WeatherReport


# App
app = FastAPI(
    title="All Networks Server",
    description="x402 server supporting Cardano, EVM, SVM, and TVM networks",
    version="2.0.0",
)


# Build accepts array dynamically based on configured addresses
accepts: list[PaymentOption] = []
if CARDANO_ADDRESS:
    accepts.append(
        PaymentOption(
            scheme="exact", pay_to=CARDANO_ADDRESS, price="$0.001", network=CARDANO_NETWORK
        )
    )
if EVM_ADDRESS:
    accepts.append(
        PaymentOption(
            scheme="exact",
            pay_to=EVM_ADDRESS,
            price="$0.001",
            network=EVM_NETWORK,
        )
    )
if SVM_ADDRESS:
    accepts.append(
        PaymentOption(
            scheme="exact",
            pay_to=SVM_ADDRESS,
            price="$0.001",
            network=SVM_NETWORK,
        )
    )
if TVM_ADDRESS:
    accepts.append(
        PaymentOption(
            scheme="exact",
            pay_to=TVM_ADDRESS,
            price="$0.001",
            network=TVM_NETWORK,
        )
    )

# x402 Middleware
facilitator = HTTPFacilitatorClient(FacilitatorConfig(url=FACILITATOR_URL))
server = x402ResourceServer(facilitator)

# Register schemes dynamically based on configured addresses
if CARDANO_ADDRESS:
    server.register(CARDANO_NETWORK, ExactCardanoServerScheme())
if EVM_ADDRESS:
    server.register(EVM_NETWORK, ExactEvmServerScheme())
if SVM_ADDRESS:
    server.register(SVM_NETWORK, ExactSvmServerScheme())
if TVM_ADDRESS:
    server.register(TVM_NETWORK, ExactTvmServerScheme())

routes = {
    "GET /weather": RouteConfig(
        accepts=accepts,
        mime_type="application/json",
        description="Weather report",
    ),
}
app.add_middleware(PaymentMiddlewareASGI, routes=routes, server=server)


# Routes
@app.get("/health")
async def health_check() -> dict[str, str]:
    return {"status": "ok"}


@app.get("/weather")
async def get_weather() -> WeatherResponse:
    return WeatherResponse(report=WeatherReport(weather="sunny", temperature=70))


if __name__ == "__main__":
    import uvicorn

    port = int(os.getenv("PORT", "4021"))
    print(f"🚀 All Networks Server listening on http://localhost:{port}")
    if CARDANO_ADDRESS:
        print(f"   Cardano: {CARDANO_ADDRESS} on {CARDANO_NETWORK}")
    if EVM_ADDRESS:
        print(f"   EVM: {EVM_ADDRESS} on {EVM_NETWORK}")
    if SVM_ADDRESS:
        print(f"   SVM: {SVM_ADDRESS} on {SVM_NETWORK}")
    if TVM_ADDRESS:
        print(f"   TVM: {TVM_ADDRESS} on {TVM_NETWORK}")
    print(f"   Facilitator: {FACILITATOR_URL}")
    print()
    uvicorn.run(app, host="0.0.0.0", port=port)
